src/librustc_parse/parser/stmt.rs

```diff
@@ -323,22 +326,83 @@ impl<'a> Parser<'a> {
         Ok((self.parse_inner_attributes()?, self.parse_block_tail(lo, blk_mode)?))
     }
 
+    fn parse_block_tail(&mut self, lo: Span, s: BlockCheckMode) -> PResult<'a, P<Block>> {
+        if self.token.is_ident() && self.look_ahead(1, |t| t == &token::Colon) {
+            // We might be having a struct literal where people forgot to include the path:
```

Comment:
Instead of introducing `parse_block_tail_inner`, could this block be extracted to a `maybe_suggest_struct_literal` or something like that and the original body of `parse_block_tail` be kept? I think it makes more sense to move the diagnostic logic into a separate function rather than the primary parsing logic.